crates/typst-ide/src/jump.rs

```diff
@@ -27,10 +31,119 @@ impl Jump {
     }
 }
 
+/// Maps a position in a document to a [jump destination][`Jump`], allowing for click-to-jump functionnality.
+pub trait JumpFromDocument {
+    type Position;
+
+    fn resolve_position(
+        &self,
+        world: &dyn IdeWorld,
+        position: &Self::Position,
+    ) -> Option<Jump>;
+}
+
+impl JumpFromDocument for PagedDocument {
+    type Position = Position;
+
+    fn resolve_position(
+        &self,
+        world: &dyn IdeWorld,
+        position: &Self::Position,
+    ) -> Option<Jump> {
+        let page = self.pages.get(position.page.get() - 1)?;
+        let click = position.point;
+
+        jump_from_click_in_frame(world, self, &page.frame, click)
+    }
+}
+
+fn nth_child(n: usize, elem: &HtmlElement) -> Option<&HtmlNode> {
+    let mut i = 0;
+    let mut was_text = false;
+    for ch in &elem.children {
+        if matches!(ch, HtmlNode::Tag(_)) {
+            continue;
+        }
+
+        let is_text = matches!(ch, HtmlNode::Text(_, _));
+        if was_text && is_text {
+            i -= 1;
+        }
+
+        if i == n {
+            return Some(ch);
+        }
+
+        was_text = is_text;
+        i += 1;
```

Comment:
the -= 1 and += 1 cancel strategy here seems slightly convoluted. I wonder whether this couldn't be simplified a bit, e.g. by moving the `i += 1` above, but skipping it in the double text case? basically only counting plus one at the start of the next child rather than the end of the current child.